Undo/COUndoTrackStore.m

```diff
@@ -190,81 +181,209 @@ - (instancetype)init
 - (void) dealloc
 {
     [_db close];
+	
+#if !(TARGET_OS_IPHONE)
+	// N.B.: We are using deployment target 10.7, so ARC does not manage libdispatch objects.
+	// If we switch to deployment target 10.8, ARC will manage libdispatch objects automatically.
+	// For GNUstep, ARC doesn't manage libdispatch objects since libobjc2 doesn't support it 
+	// currently (we compile CoreObject with -DOS_OBJECT_USE_OBJC=0).
+	dispatch_release(_queue);
+	dispatch_release(_transactionLock);
+#endif
+}
+
+- (BOOL) setupSchema
+{
+    /* SQLite compatibility */
+    
+	[_db executeUpdate: @"PRAGMA foreign_keys = ON"];
+	if (1 != [_db intForQuery: @"PRAGMA foreign_keys"])
+	{
+		[NSException raise: NSGenericException
+		            format: @"Your SQLite version doesn't support foreign keys"];
+	}
+	
+	[_db beginDeferredTransaction];
+
+	/* Store Metadata table (including schema version) */
+
+	if (![_db tableExists: @"storeMetadata"])
+	{
+		NSAssert(![_db tableExists: @"commands"], @"Unsupported unversioned schema");
+
+		[_db executeUpdate: @"CREATE TABLE storeMetadata(version INTEGER)"];
+		[_db executeUpdate: @"INSERT INTO storeMetadata VALUES(1)"];
+	}
+	else
+	{
+		int version = [_db intForQuery: @"SELECT version FROM storeMetadata"];
+		if (1 != version)
+		{
+			NSLog(@"Error, undo track store version %d, only version 1 is supported", version);
+			[_db rollback];
+			return NO;
+		}
+	}
+	
+	/* Commands and Tracks tables */
+	
+	[_db executeUpdate: @"CREATE TABLE IF NOT EXISTS commands (id INTEGER PRIMARY KEY AUTOINCREMENT, "
+	                     "uuid BLOB NOT NULL UNIQUE, parentid INTEGER, trackname STRING NOT NULL, data BLOB NOT NULL, "
+	                     "metadata BLOB, timestamp INTEGER NOT NULL, deleted BOOLEAN DEFAULT 0)"];
+	
+	// NULL currentid means "the start of the track"
+	[_db executeUpdate: @"CREATE TABLE IF NOT EXISTS tracks (trackname STRING PRIMARY KEY, "
+	                     "headid INTEGER NOT NULL, currentid INTEGER, "
+	                     "FOREIGN KEY(headid) REFERENCES commands(id), "
+	                     "FOREIGN KEY(currentid) REFERENCES commands(id))"];
+
+    [_db commit];
+    
+    if ([_db hadError])
+    {
+        NSLog(@"Error %d: %@", [_db lastErrorCode], [_db lastErrorMessage]);
+        return NO;
+    }
+    return YES;
 }
 
 - (BOOL) beginTransaction
 {
-    return [_db executeUpdate: @"BEGIN EXCLUSIVE TRANSACTION"];
+	ETAssert([NSThread isMainThread]);
+
+	// If there is a background operation (e.g. mark as deleted, vacuum) underway,
+	// wait until it is finished
+	dispatch_semaphore_wait(_transactionLock, DISPATCH_TIME_FOREVER);
+
+    BOOL ok = [_db beginTransaction];
+	if (!ok)
+	{
+		dispatch_semaphore_signal(_transactionLock);
+	}
+	return ok;
 }
 
 - (BOOL) commitTransaction
 {
+	ETAssert([NSThread isMainThread]);
+
     BOOL ok = [_db commit];
```

Comment:
-commit sends a regular SQL statement, it should be done in the `_queue`, I think.
While SQLIte is supposed to be threadsafe by default (see https://www.sqlite.org/threadsafe.html ) I don't think fmdb is.